### PR TITLE
Add unit tests for drop_lsn and fix.

### PR DESCRIPTION
**Pull request by lubennikovaav**
test_drop_extend and test_truncate_extend illustrate what happens if we dropped a segment and then created it again within the same layer.
Fix: if get_layer_for_write() finds an existing open layer that is dropped, just mark it as non-writeable and create a new open layer for write.

TODO:
- [x] handle non-writeable open layers in checkpointer.
- [x] remove code that handles dropped layers in freeze() function. It is not used anymore.

### CODE REVIEW

**Review by hlinnaka — COMMENTED** [review]
Makes sense. This will probably conflict with https://github.com/zenithdb/zenith/pull/572.

**Review by patins — COMMENTED** [review]
When a segment is dropped and then reused later (before checkpointing occurs) I think this should fail at the `SegEntry::insert_open` assert. How do we ensure dropped layers ever get checkpointed?

File: pageserver/src/layered_repository/layer_map.rs [hunk 1]
```diff
@@ -295,7 +295,10 @@ impl SegEntry {
     }
 
     pub fn insert_open(&mut self, layer: Arc<InMemoryLayer>) {
-        assert!(self.open.is_none());
+        if let Some(prev_open) = &self.open {
+            assert!(!prev_open.is_writeable());
+            println!("prev_open exists. non writeable");
```
[the remark this answers is not in the document]
Comment (lubennikovaav) [reply pageserver/src/layered_repository/layer_map.rs]: Fixed

File: pageserver/src/layered_repository.rs [hunk 2]
```diff
@@ -1165,7 +1165,14 @@ impl LayeredTimeline {
             if layer.get_start_lsn() > lsn {
                 bail!("unexpected open layer in the future");
             }
-            return Ok(layer);
+
+            if layer.is_dropped() {
+                layer.mark_non_writeable();
+            }
```
[the remark this answers is not in the document]
Comment (lubennikovaav) [reply pageserver/src/layered_repository.rs]: Fixed

File: pageserver/src/layered_repository.rs [hunk 3]
```diff
@@ -1300,27 +1320,53 @@ impl LayeredTimeline {
...
+                layers.pop_oldest_open();
+            } else {
+                let FreezeLayers {
+                    frozen: frozen_result,
+                    open: open_result,
+                } = oldest_layer.freeze(last_record_lsn)?;
+
+                frozen = frozen_result;
+                maybe_new_open = open_result;
+
```
Comment (lubennikovaav) [inline pageserver/src/layered_repository.rs:1349]: @patins , is there some better rust syntax for this?

File: pageserver/src/layered_repository.rs [hunk 4]
```diff
@@ -1300,27 +1320,53 @@ impl LayeredTimeline {
...
             // Write the now-frozen layer to disk. That could take a while, so release the lock while do it
             drop(layers);
             let new_historics = frozen.write_to_disk(self)?;
-            layers = self.layers.lock().unwrap();
 
             // Finally, replace the frozen in-memory layer with the new on-disk layers
+            layers = self.layers.lock().unwrap();
             layers.remove_historic(frozen.as_ref());
 
+            //FIXME This needs a comment.
```
Comment (lubennikovaav) [inline pageserver/src/layered_repository.rs:1369]: @patins , could you please add a comment for this code introduced by your previous commit.
Feel free to push to this branch.

File: pageserver/src/layered_repository/layer_map.rs [hunk 5]
```diff
@@ -106,11 +106,13 @@ impl LayerMap {
...
-            segentry.open.as_ref().unwrap(),
-            &oldest_entry.layer
-        ));
-        segentry.open = None;
+        if Arc::ptr_eq(segentry.open.as_ref().unwrap(), &oldest_entry.layer) {
+            segentry.open = None;
+        } else {
+            // We could have already updated segentry.open for
+            // dropped (non-writeable) layer. This is fine.
+            assert!(!&oldest_entry.layer.is_writeable());
```
Comment (patins) [inline pageserver/src/layered_repository/layer_map.rs:114]: Is the & needed? Also lets add assert dropped

File: pageserver/src/layered_repository.rs (hunk 3 above)
Comment (patins) [reply pageserver/src/layered_repository.rs:1349]: Nothing that @sharnoff or I could come up with. Would probably change it to
```
let freeze_layers = oldest_layer.freeze(...)?;
frozen = freeze_layers.frozen;
maybe_new_open = freeze_layers.open;
```

File: pageserver/src/layered_repository/layer_map.rs [hunk 6]
```diff
@@ -85,7 +85,7 @@ impl LayerMap {
     pub fn insert_open(&mut self, layer: Arc<InMemoryLayer>) {
```
[the remark this answers is not in the document]
Comment (lubennikovaav) [reply pageserver/src/layered_repository/layer_map.rs:85]: I didn't change it intentionally, because this function actually inserts the given layer into the `open_layers` list.

File: pageserver/src/layered_repository.rs [hunk 7]
```diff
@@ -1300,27 +1320,53 @@ impl LayeredTimeline {
...
-
-            // replace this layer with the new layers that 'freeze' returned
-            layers.pop_oldest_open();
-            if let Some(new_open) = maybe_new_open.clone() {
-                layers.insert_open(new_open);
+            //
+            let frozen: Arc<InMemoryLayer>;
+            let maybe_new_open: Option<Arc<InMemoryLayer>>;
+
+            if !oldest_layer.is_writeable() {
```
[the remark this answers is not in the document]
Comment (lubennikovaav) [reply pageserver/src/layered_repository.rs]: Good catch. I moved this check into freeze().

File: pageserver/src/layered_repository/layer_map.rs [hunk 8]
```diff
@@ -294,8 +297,13 @@ impl SegEntry {
...
-    pub fn insert_open(&mut self, layer: Arc<InMemoryLayer>) {
-        assert!(self.open.is_none());
+    // Set new open layer for a SegEntry.
+    // It's ok to rewrite previous open layer,
+    // but only if it is not writeable anymore.
+    pub fn update_open(&mut self, layer: Arc<InMemoryLayer>) {
+        if let Some(prev_open) = &self.open {
+            assert!(!prev_open.is_writeable());
+        }
         self.open = Some(layer);
```
[the remark this answers is not in the document]
Comment (lubennikovaav) [reply pageserver/src/layered_repository/layer_map.rs:307]: > Allowing replacing the open layer seems a bit sloppy to me. Previously, it was clear that each segment would have only one open layer at a time, now the rule is more complicated.

Do you see any issues with correctness here?
We should definitely use better terminology here because now it is quite confusing.
I think if we rename `open_layers` to `inmemory_layers`, things will be more clear - multiple in-memory layers can exist, but only one of them is open (aka writeable).
But I'd rather do this as a separate PR.

[the remark this answers is not in the document]
Comment (hlinnaka) [reply pageserver/src/layered_repository/layer_map.rs:307]: > > Allowing replacing the open layer seems a bit sloppy to me. Previously, it was clear that each segment would have only one open layer at a time, now the rule is more complicated.
> 
> Do you see any issues with correctness here?

No. It's just a bit scary.

> We should definitely use better terminology here because now it is quite confusing.
> I think if we rename `open_layers` to `inmemory_layers`, things will be more clear - multiple in-memory layers can exist, but only one of them is open (aka writeable).
> But I'd rather do this as a separate PR.

Ok, works for me.

File: pageserver/src/layered_repository.rs [hunk 9]
```diff
@@ -1326,17 +1347,19 @@ impl LayeredTimeline {
...
             // the frozen layer to disk, and replace the in-memory frozen layer
             // with the new on-disk layers.
-            let FreezeLayers {
-                frozen,
-                open: maybe_new_open,
-            } = oldest_layer.freeze(last_record_lsn)?;
+            //
+            let freeze_layers = oldest_layer.freeze(last_record_lsn)?;
+            let frozen = freeze_layers.frozen;
+            let maybe_new_open = freeze_layers.open;
```
Comment (patins) [inline pageserver/src/layered_repository.rs:1353]: Was this change intentional?

**Review by patins — APPROVED** [review]

File: pageserver/src/layered_repository.rs (hunk 9 above)
Comment (lubennikovaav) [reply pageserver/src/layered_repository.rs]: It was at first but became unneeded after further refactoring.
